### ISSUE TITLE
Code Review comments

### ISSUE THREAD

**Issue by bpaauwe** [sjpbailey/udi-poly-baspi6u6r-garage-door-master#2]
I did a quick review of the node server and found the following (mostly minor) issues:

in GarageDoorsController.py
  - In init() what is 'self.Device = Device' for?  I don't see self.Device being used anywhere.
  - in start() why are you calling add_custom_config_docs() with sample data?  This should probably be removed since you have a node server configuration file.
  - What is the class bc being used for?  
  - process_config() appears to be an unused function
  - set_module_logs() appears to be an unused function
  - in check_params() the variable 'st' is not used for anything
  - The function remove_notice_test() and the associated command are an example from the template, not something that you should have in your node server.

in GarageDoorsNodes.py
  - you pass in both the ip address and the device object (bc) and then re-allocate the bc object in start(). That seems unnecessary. 
  - Can self.bc.ePlatform actually have multiple values?  If not, you should be using and if / else construct to determine which value it is currently set to:
     if self.bc.ePlatform == Platform.BASC_NONE:
     elif self.bc.ePlatform == Platform.BASC_PI:
     elif self.bc.ePlatform == Platform.BASC_ED:
    That will be more Python like, readable, and more efficient.
  - loops and arrays would make the code more readable and simplify things quite a bit. For example where you have the log output for the 6 universal inputs
```Python
      for i in range(1,7):
           LOGGER.info(self.bc.univeralInput(i))
```
  
  - There's a lot of nearly duplicate code. This will make it much harder to maintain and the likelyhood of missing something when/if you have to update this is much higher.  
  
For example, you could make a function:

```Python
def` setInputDriver(self, driver, input):
    input_val = self.bc.universalInput(intput)
    count = 0
    if input_val is not None:
        count = int(float(input_val))/1000
        self.setDriver(driver, count, force=True)
```

And then in start do:
```Python
    self.setInputDriver('GV6', 1)
    self.setInputDriver('GV7', 2) 
    self.setInputDriver('GV8', 3)
    self.setInputDriver('GV9', 1)
    self.setInputDriver('GV10', 2) 
    self.setInputDriver('GV11', 3)
````

The logic for the setOn functions is confusing to me.  It seems to be:
   if output is off, turn it on, wait, turn it off
   if output is on turn it off

Same with the bump functions, they behave differently if the initial state is on vs. if it is off.  If this is the correct logic, then it would be helpful to document it (and the reason for it) so that others can better understand the code.

And as I mentioned above, each of these setOn and Bump functions are basically the same so it would better to just use one function for each instead of 6.  I'd create an array to map the input and output drivers to the intput/output number and then use the command to determine which input/output is being controlled.  Something like:

```Python
self.mapping = {
    'BON1' : {'intput':'GV6', 'output':'GV0', 'index': 1}, 
    'BON2' : {'intput':'GV7', 'output':'GV1', 'index': 2}, 
    'BON3' : {'intput':'GV8', 'output':'GV2', 'index': 3}, 
    'BON4' : {'intput':'GV9', 'output':'GV3', 'index': 4}, 
    'BON5' : {'intput':'GV10', 'output':'GV4', 'index': 5}, 
    'BON6' : {'intput':'GV11', 'output':'GV5', 'index': 6},
    'BOB1' : {'intput':'GV6', 'output':'GV0', 'index': 1}, 
    'BOB2' : {'intput':'GV7', 'output':'GV1', 'index': 2}, 
    'BOB3' : {'intput':'GV8', 'output':'GV2', 'index': 3}, 
    'BOB4' : {'intput':'GV9', 'output':'GV3', 'index': 4}, 
    'BOB5' : {'intput':'GV10', 'output':'GV4', 'index': 5}, 
    'BOB6' : {'intput':'GV11', 'output':'GV5', 'index': 6},
}

def setOn(self, command):
    input = self.mapping[command]['input']
    output = self.mapping[command]['output']
    index = self.mapping[command]['index']
    if binaryoutput(index) != 1:
        binaryOutput(index, 1)
        self.setDriver(input, 1)
        self.setDriver(output, 255)
        LOGGER.info('Output {} On'.format(index))
        time.sleep(2)
    if binaryoutput(index) != 0:
        binaryoutput(index, 0)
        self.setDriver(input, 0)
        LOGGER.info('Output {} Off'.format(index))
        time.sleep(15)
        self.doorstat(index, output)
```

**Comment by sjpbailey** [comment]
Thank you @Bob 
Very helpful to a newbie!

**Comment by sjpbailey** [comment]
@bob 
I will clean this up and resubmit it after some more testing! 
I really appreciate your input it is very, very helpful as I am learning many new things!
Thank You I will get back to you after I clean up my inexperience.

**Comment by bpaauwe** [comment]
You are welcome.  I wish more people would review the node servers code before we approve it.  [yes, even mine].  It would improve the node servers and provide valuable learning (both for the reviewer and the reviewee)

This will become more important when we move to PG3 with non-free node servers.

Since you're going to re-submit a new version, can I go ahead and deny the current submission?  I'm actually curious what happens when one is denied, I think anyone has ever done that :)

**Comment by sjpbailey** [comment]
I also need to add the CAD drawing I did I to the readme then I can clean my commits and get one in as clean as can be lol!
Thank You again as you know there is always more than one way to peal a potato!

**Comment by sjpbailey** [comment]
PS
 I have a new module for this CC stuff I need to get on PYPI. 
It uses an internal CC Sedona program. 
I am finishing a thermostat or fan coil.
I am just about done with that so I can install it on my split HVAC system in the attic. 
The poly will override 4 of the 24 VIRTUAL points I can now read and write to the Device.
The four points or nodes are to send setpoint and overrides schedule to the Sedona program.
I am also going to add a dump damper on the supply air to give us a whole house fan.
Have a great Day Bob!

**Comment by bpaauwe** [comment]
> Sure @bpaauwe Bob do it!
> 
> I like your ideas to clean it up with single functions and want to try it and see if I can get it condensed.
> I was adding the confusing open close bump separate to insure if 6 doors are installed not all 6 can operate at once.

Making them all separate to start is a good process to get it working, but take the time to look over your code for areas where you have duplicate (or near duplicate) sections of code. Pulling that out to a function and getting rid of the duplication is a good programming practice.  It makes it less bug prone when there is just one place to fix a bug. (I.E. if you fix one section but forget to fix the other). 

> There is only one contact closure to open and close the garage so I check if it if not on turn it on for time.sleep(2) emulating a button push. then it has to positively turn off.

I understand.  It's just that the logic in the program is a bit hard to follow.  Since you're trying to simulate a button press with an open/close contact, maybe create a function called 'button_press' that does the close/wait/open sequence (take the output as a parameter) and then call that to when you need a button press.

> one thing it is lacking with the delays is that it can retriever after the 15 second delay if pushed twice so I want to see if I can positively lock out any commands while the door is operating.

I watched your video on it.  It seemed like there were a lot of delays in updating status.  I don't know anything about the library you're using to interface with I/O board, if it has some way to register a callback to receive status events you could use that to make status updates immediate.  If not, maybe you could create a thread that polls for the updates at a fairly rapid rate (1 second or so).

You can use a global flag (or flags) to track when a door is opening or closing and then ignore or delay new commands if that flag is set.  But the user may want to stop it partially open or close and, at least with my opener, pressing the button will stop the current operation at whatever position the door is at.

> All safeties are on the door and in playing with it I have not clobbered myself as the IR door sensor is the thing that protects everything.
That good!

**Comment by sjpbailey** [comment]
Fantastic!

 I have made most all of your suggested adjustments. 
Currently working on re-working control into one, instead of six functions inputs are done. 
I still need six profile pushbuttons for each door in AC so I am working that out, mainly in my brain lol! 

 Immediate Status of action has been a problem for me and to see a work around is great with one second intervals.

 
 Future:
This is why I am now doing function in the Device itself with Contemporary Controls programming tool called Sedona. 
 
 Now all my physical nodes will report the same however function will be influenced via the up to 24 Virtual points each device has. So physical points or nodes binaryOutput/universalInput BASpi-6U6R has six in six out 24 virtual in/out's. 
 So ISY program will not be the main controller the devices physical points.
 
 Pretty interesting stuff and it will have tighter control (PID) say for the fan coil unit I am working on. 
 Controlling these Virtual Nodes: Occupied/un-Occupied, Fan Override, Heat Setpoint, Cool Setpoint, Heat Off Cool & Whole house fan with a dump damper.  
So now I am the widget creator in BACnet and then polyglot  to see it as a node.

Never ever thought I could do this!
Thank You Bob!

**Comment by sjpbailey** [comment]
Hello Bob,

I dislike to bug anyone with my questions however you seem to know what I am trying to do and have been a huge help cleaning up my hacks.
 If you don't mind.

Status of progress.

1. - in GarageDoorsController.py
- In init() what is 'self.Device = Device' for? I don't see self.Device being used anywhere.
- in start() why are you calling add_custom_config_docs() with sample data? This should probably be removed since you have a node server configuration file.
- What is the class bc being used for?
- process_config() appears to be an unused function
- set_module_logs() appears to be an unused function
- in check_params() the variable 'st' is not used for anything
- The function remove_notice_test() and the associated command are an example from the template, not something that you should have in your node server.

All Done.

2. in GarageDoorsNodes.py

- For start:

````
def start(self):
        ### Do we have a BASpi or an Edge Device ###
        if self.bc.ePlatform == Platform.BASC_PI: ### if a BASpi-6u6r Device is found
            LOGGER.info('connected to Garage Door 1-6')
        elif self.bc.ePlatform == Platform.BASC_ED: ### if a BASpi-Edge Device is found
            LOGGER.info('connected to BASpi-Edge Module ONE')
            self.setDriver('ST', 1)    
        elif self.bc.ePlatform == Platform.BASC_NONE: ### if there is no BASpi Device found
            LOGGER.info('Unable to connect to Garage Doo 1-6')
            LOGGER.info('ipaddress')
        else:
            pass
```

- For  Logger and Inputs       

       ```
 # How many nodes or points does the device have
        LOGGER.info('\t' + str(self.bc.uiQty) + ' Universal inputs in this Doors 1-6')
        LOGGER.info('\t' + str(self.bc.boQty) + ' Binary outputs in this Doors 1-6')
        LOGGER.info('\t' + str(self.bc.vtQty) + ' Virtual points In This Doors 1-6')
        
        # What are the Physical Nodes or Device point currently reading
        LOGGER.info('Inputs')
        for i in range(1,7):
               LOGGER.info(str(self.bc.universalInput(i)))
        LOGGER.info('Outputs')
        for i in range(1,7):
               LOGGER.info(str(self.bc.binaryOutput(i)))  `

### Universal Inputs
        self.setInputDriver('GV0', 1)
        self.setInputDriver('GV1', 2)
        self.setInputDriver('GV2', 3)
        self.setInputDriver('GV3', 4)
        self.setInputDriver('GV4', 5)
        self.setInputDriver('GV5', 6)
       
    ### Universal Inputs Conversion ###
    def setInputDriver(self, driver, input):
        input_val = self.bc.universalInput(input)
        count = 0
        if input_val is not None:
            count = int(float(input_val))//1000
            self.setDriver(driver, count, force=True)
```
 
 So I also fixed the trigger and retriever so you can stop the Door anywhere you want just like the button.
 
```
 ### Output Door-1
    def setOn(self,command=None):
        GV6 = command.get('value')
        if self.bc.binaryOutput(1) != 1:
            self.bc.binaryOutput(1, 1)
            self.setDriver("GV6", 1)
            self.setDriver("GV0", 255)
            LOGGER.info('Output 1 On')
            time.sleep(1)
            self.bc.binaryOutput(1, 0)
            self.setDriver("GV6", 0)
            self.setDriver("GV0", 255)
            LOGGER.info('Output 1 Off')
```

- Okay now for feedback Status for the setOn function above. 
- 
- I tried but need to learn more on dict structure as I was getting unhasable 'dict' errors mapping function while trying to condense this function.
- 
- `input = self.mapping[command]['input']`
- TypeError: unhashable type: 'dict'
- 
- So I went to a toggle only and ignored status of the commands and now a push will stop the Garage Door anywhere you want in its travel.
- 
- 
-  Now looking at door one, the door gets the setOn function command and it takes 15 seconds to cycle open or close.
-  So for the universalInput status using a resistor and from NLS show open, stopped and closed based on 12,000 ohm closed, 0 ohm stopped and 5000 ohm open.
-  
-   Now if I do nothing in the GarageDoorsNodes.py  I will get the feedback I desire when the poly is short poll refreshed?
-  Now how can I read that input converted to a driver GV0 to read on a change of state? 
-  Now I could add a virtual point into the Device and have it report as a universalInput or a binaryOutput but that also does not toggle unless forced to within a function or on a refresh?
-  
-  So if I delay and wait, I stop my script and the Garage Door cannot be re-triggered to stop it anywhere you want, yuck.
-  
-  If I go through the effort of mapping a resistor ladder how come I cannot read it any time it changes state and if I can every second how do I do that. 
-  Also for input output controller for commands, how do I separate them into one function when they are separate nodes that have to have a separate command to all 6?

-  So everything is golden except the feedback and I cannot get it into my head no how that should work.
-  I get lost with all the magic and my dyslexia sometimes!
-  
-  I hope this all makes sense because I do not speak English or Python very well.
-  
-  Highest Regards to You @bpaauwe

**Comment by bpaauwe** [comment]
>  ### Output Door-1
>     def setOn(self,command=None):
>         GV6 = command.get('value')
>         if self.bc.binaryOutput(1) != 1:
>             self.bc.binaryOutput(1, 1)
>             self.setDriver("GV6", 1)
>             self.setDriver("GV0", 255)
>             LOGGER.info('Output 1 On')
>             time.sleep(1)
>             self.bc.binaryOutput(1, 0)
>             self.setDriver("GV6", 0)
>             self.setDriver("GV0", 255)
>             LOGGER.info('Output 1 Off')
> ```
For the above function you need to convert or map the incoming command value to the input/output driver strings.  I'm not sure what you actually get from 'command.get()' but the idea is to use that as the key into a dict that holds the driver strings.

Somewhere above you'd have to create the dict with fixed values (since the input and output driver strings are fixed, known entities).  Then in the setDriver() calls, you use use the driver strings in the dict instead of the literals you have now.

If you have 6 commands defined, say button1 ... button6. You can map all 6 of those to the same command handler in your code. The handler will be called with the name of the button that was invoked and any values associated with that.  So in your handler you convert the name to a driver string via a mapping dictionary.
```Python
button_map = { 
    'button1': { 'in_driver': 'GV0', 'out_driver': 'GV6' },
    'button2': { 'in_driver': 'GV1', 'out_driver': 'GV7' },
    'button3': { 'in_driver': 'GV2', 'out_driver': 'GV8' },
    'button4': { 'in_driver': 'GV3', 'out_driver': 'GV9' },
    'button5': { 'in_driver': 'GV4', 'out_driver': 'GV10' },
    'button6': { 'in_driver': 'GV5', 'out_driver': 'GV11' },
    }

#  button_map['button1']['in_driver']  is the driver string "GV0"
self.setDriver(button_map[cmd]['in_driver], 0)
self.setDriver(button_map[cmd]['out_driver], 255)
```
Does that make more sense?

> -  If I go through the effort of mapping a resistor ladder how come I cannot read it any time it changes state and if I can every second how do I do that. 

If I understand the API for the BASpi devices, it looks like you have to query them to get the status.  I can't find any documentation on-line for the API so I'm just kind of guessing based on your control module.

In that case, you'd need to implement a thread to watch for, and report status changes.  This may be something you'd want to implement in your controller module instead of in the node server.  But the general idea would be set up an command queue and a main loop.

The general idea would be something like:

```Python
while True:
    if input_queue has command:
        if hardware command:
            send command to hardware
            remove command from queue
        else if command is STOP:
            print('stopping command/status loop')
            break

    for each output:
        query output
        if output status has changed:
            Notify any listeners of output status change

    # add a sleep here, how long depends on how much of a load
    # you want to inflict on the system and what latency you can live with.
    # 1 second might be a good starting point.
```

> -  Also for input output controller for commands, how do I separate them into one function when they are separate nodes that have to have a separate command to all 6?

I'm not sure I understand the question.  If you have separate nodes, then each node is represented a separate node object. The node class would have one function that handled commands. Each node object would have its own instance of the function, but you'd only write the function once.

**Comment by sjpbailey** [comment]
I took a needed break to CAD my son's pool house this weekend, lol!

Yes direct to the module no API, the EDGE is full IOT with a cloud, I have not even looked at. 

Very well that give me a starting point, hopefully get this done and re-summited it.

It will be a little while I will let you know when done!

You Are Very Much Appreciated For Your Time!

Thank You @bpaauwe

**Comment by sjpbailey** [comment]
DOH!

I need to edit my module bascontrolns.py for readObject. 

I am so glad you have patients with me I am feel like one dull tool in the shed.

```
###########################################################################
    #
    def readObject(self , iIndex):

        """Returns response text if success, None if failure"""

        # If not online.
        if not self.online:
            # If retry period has expired.
            if int(time.time()) >= self.nextRetry:
                # If we're still offline.
                if not self.isOnline():
                    # New retry time has been set.
                    return None
            # Else need to wait.
            else:
                return None

        # Assume failure.
        retval = None
        # XML read request string.
        req = '<rdom fcn="get" doc="rtd" path="obj[' + str(iIndex) + ']/val_scl"/>'
        # Try to execute the POST.
        try:
            rsp = requests.post(self.sFullUrl, data=req , timeout=self.timeout)
            # If good HTML response.
            if (rsp.status_code == 200):
                root = et.fromstring(rsp.text)
                if root.attrib['rsp'] == 'ack':
                    retval = root.text
            else:
                # Something there, but not BAScontrol.
                self.online = False
        except OSError:
            # Timeout.
            self.online = False
        # If we were taken offline.
        if not self.online:
            # Set for retry.
            self.nextRetry = int(time.time())+self.retryInterval
            # Raise an exception.
            ###raise IOError
            # NOT REACHED
        # And return.
        return retval 
```

**Comment by sjpbailey** [comment]
@bpaauwe,

Hello Bob,
  I had a break and still need to finish my oldest plans for their pool house in CAD and am getting back to this have cleared my commits twice probably up to 500 or so.
 
 Your explanation for mapping the commands was already understood. I guess I did not know how to ask and probably still do not, here is what I wanted which is:
 
 How do I write the syntax to call the command which are the keys 'BON1' - 'BON6' to the mapped values in the dictionary.
 
 This works every button opens door one:
 `    self.mapping = {
                'BON1' : {'input':'GV6', 'output':'GV0', 'index': (1)}, 
                'BON2' : {'input':'GV7', 'output':'GV1', 'index': (2)}, 
                'BON3' : {'input':'GV8', 'output':'GV2', 'index': (3)}, 
                'BON4' : {'input':'GV9', 'output':'GV3', 'index': (4)}, 
                'BON5' : {'input':'GV10', 'output':'GV4', 'index': (5)}, 
                'BON6' : {'input':'GV11', 'output':'GV5', 'index': (6)},
                } `
 `def setOn(self, command):
        input = self.mapping['BON1']['input']
        output = self.mapping['BON1']['output']
        index = self.mapping['BON1']['index']
        if self.bc.binaryOutput(index) == 0:
            self.bc.binaryOutput(index, 1)
            self.setDriver(input, 1)
            self.setDriver(output, 255)
            LOGGER.info('Output {} On'.format(index))
            time.sleep(1)
        if self.bc.binaryOutput(index) != 0:
            self.bc.binaryOutput(index, 0)
            self.setDriver(input, 0)
            LOGGER.info('Output {} Off'.format(index))`
   
   How I am send out to profiles the commands setOn as BON1-BON6 and QUERY.      
            ` commands = {
                    'BON1': setOn,
                    'BON2': setOn,
                    'BON3': setOn,
                    'BON4': setOn,
                    'BON5': setOn,
                    'BON6': setOn,
                    'QUERY': query,
                    
                }`

This does not work:
 
 `def setOn(self, command):
        input = self.mapping[command]['input']
        output = self.mapping[command]['output']
        index = self.mapping[command]['index']`
        
 I have tried many many syntax and read through all of my errors trying command as tuple, int, str dead ends all.
Now when I give it a single key 'BON1' above all buttons naturally work. 
Also added how I am send out the commands setOn as BON1-BON6 and QUERY.

Part two status:

 Have not gotten anywhere editing my module bascontrolns https://github.com/sjpbailey/bascontrol_ns on a command call, YET.
 Testing with a local copy bascontrolns_test.

 Now the thing is the node server sends the command out it then takes 12 seconds to fully open and close the door. 
Now my module already sent its values when asked and is waiting to be asked again.
If a short poll in the controller updates my drivers also my status why can I not call that every 12 seconds in a for loop in my node server.
 The module will sending an update will not be seen by the nod server because it has not asked for it.
  
 In other words I can window into the controller completely separate from everything in its own GUI that lets me override the node values sent to the node server. 
 If I act like a Module sending out from the GUI just like a for loop with changed values in the bascontrolns module it will do nothing in the node server, because it has not refreshed to see a change.  
 
 Now I can live with it showing OPERATING and waiting for a proper status OPEN or CLOSE on a refresh.
 Now what if I am a temperature waiting to turn on a fan in an ISY AC program? 
 Do I over or under shoot my target setpoint because I had to wait for the node server to send me updated values? 
 We are the API in an RPi as the BASpi not running its own sequence to be called, it just exposes its nodes.

Lastly, It seems to work with time.sleep however I do not want to time.sleep to wait for a status anymore because I want to be able to stop the door anywhere which you wisely pointed out.

Sorry for the repetitive question taking up your time! 
 
 Highest Regards

**Comment by bpaauwe** [comment]
> How do I write the syntax to call the command which are the keys 'BON1' - 'BON6' to the mapped values in the dictionary.
> 
>    How I send out to profiles the commands setOn as BON1-BON6 and QUERY.      
>              ```
> commands = {
>                     'BON1': setOn,
>                     'BON2': setOn,
>                     'BON3': setOn,
>                     'BON4': setOn,
>                     'BON5': setOn,
>                     'BON6': setOn,
>                     'QUERY': query,
>                     
>                 }
> ```
> 
> This does not work:
> 
> ```
> def setOn(self, command):
>        input = self.mapping[command]['input']
>        output = self.mapping[command]['output']
>        index = self.mapping[command]['index']
>        
> ```

You're very close.  command is an object (a dict, I believe).  You'll need to reference the correct key from that to get the information needed for the mapping.  I don't remember the exact structure of the dict so do a:

``` Python
LOGGER.error('command = {}'.format(command))
```
in setOn() to see what it looks like.


> Part two status:
> 
> Have not gotten anywhere editing my module bascontrolns https://github.com/sjpbailey/bascontrol_ns on a command call, YET.
> Testing with a local copy bascontrolns_test.
> 
> Now the thing is the node server sends the command out it then takes 12 seconds to fully open and close the door.
> Now my module already sent its values when asked and is waiting to be asked again.
> If a short poll in the controller updates my drivers also my status why can I not call that every 12 seconds in a for loop in my node server.
> The module will sending an update will not be seen by the nod server because it has not asked for it.

Your node server can do whatever you want it to.  Its role is to act as a translator and translate commands and status between the physical device and the ISY.  Don't get hung up on what the template node server does, it's just an example to show how to use the functions in the interface module.

The polyinterface module's poll function is an optional helper function that can be used to trigger your node server to do something on a regular interval.  This is handy if you have a device that you only need to query for current status periodically. But it's not very useful when you are interactively controlling a device.

Your garage door control is not something that updates its status every 12 seconds or every 5 minutes.  Events can happen at any time that invoke a status change and the node server should be responding to them when they happen and update the ISY immediately, not wait for a poll interval.

The flow should be something like this:
1. ISY sends command to node server
2. node server sends command to [your bascontrolns module?]
3. either the bascontrolns modules notifies the node server of a status update or the node server queries the bascontrolns module for the current status
4. the node server sets the driver to the current status

There's no polling involved in that flow.

If a remote is used to operate the garage door, you need a way to notify the ISY of the status changes when this happens.  I don't know anything about the bascontrolns module.  Ideally, it would have a way to notify the node server of the status change directly.  But if not, then here is where you'd want to poll and you'd have to live with the delays.  

Typically, with a hardware device that has inputs like this, the hardware would raise an interrupt when the input line changed state.  That would cause an interrupt handler routine to examine the input lines to see what had changed.  This would then trigger a thread/process to notify anyone interested in that info that it changed.  That could be calling a registered callback function or sending a request over a network socket or sending a request to a web server.  Since I don't really know anything about the hardware or the bascontrolns module, I can't be more specific and this is just generally how things like this are handled.

I'm not sure I can provide much more without knowing a lot more about the hardware and software interfaces for the hardware.

**Comment by sjpbailey** [comment]
Hello @bpaauwe,

 Well that did it found the keys: 
```
 LOGGER.error('command = {}'.format(command))
            LOGGER.error(command.values()) #(['baspidoor1_id', 'BON1', {}])
            LOGGER.error(command.keys()) #(['address', 'cmd', 'query']) 
```
 
 Also in the poly interface:
```
 def runCmd(self, command):
        if command['cmd'] in self.commands:
            fun = self.commands[command['cmd']]
            fun(self, command)
```

Commands are now:
``````
# Input Output Control       

        self.mapping = {
            'BON1' : {'input':'GV6', 'output':'GV0', 'index': (1)}, 
            'BON2' : {'input':'GV7', 'output':'GV1', 'index': (2)}, 
            'BON3' : {'input':'GV8', 'output':'GV2', 'index': (3)}, 
            'BON4' : {'input':'GV9', 'output':'GV3', 'index': (4)}, 
            'BON5' : {'input':'GV10', 'output':'GV4', 'index': (5)}, 
            'BON6' : {'input':'GV11', 'output':'GV5', 'index': (6)},
            'BOB1' : {'input':'GV6', 'output':'GV0', 'index': (1)}, 
            'BOB2' : {'input':'GV7', 'output':'GV1', 'index': (2)}, 
            'BOB3' : {'input':'GV8', 'output':'GV2', 'index': (3)}, 
            'BOB4' : {'input':'GV9', 'output':'GV3', 'index': (4)}, 
            'BOB5' : {'input':'GV10', 'output':'GV4', 'index': (5)}, 
            'BOB6' : {'input':'GV11', 'output':'GV5', 'index': (6)},
            } 
    
    # Used to show correct status with time delay between commands of 15 seconds
    # Door will travel then status will show after a 15 second time.sleep(15) 
    def setOn(self, command):
        input = self.mapping[command['cmd']]['input']
        output = self.mapping[command['cmd']]['output']
        index = self.mapping[command['cmd']]['index']
        if self.bc.binaryOutput(index) == 0:
            self.bc.binaryOutput(index, 1)
            self.setDriver(input, 1)
            self.setDriver(output, 255)
            LOGGER.info('Output {} On'.format(index))
            time.sleep(1)
        if self.bc.binaryOutput(index) != 0:
            self.bc.binaryOutput(index, 0)
            self.setDriver(input, 0)
            LOGGER.info('Output {} Off'.format(index))
            LOGGER.info('Door Operating')
            time.sleep(12)
        if self.bc.universalInput(index) == 12:
            self.setDrivers(output, '12', report=True, force=True)
        if self.bc.universalInput(index) == 5:
            self.setDriver(output, '5', report=True, force=True)
        if self.bc.universalInput(index) == 0:
            self.setDriver(output, '0', report=True, force=True)
            LOGGER.info('Door Operation Complete')
       
    # Used to show STOPPED status with no time delay between commands so you can pposition door or stop it anywhere
    # Shows status of STOPPED if door has traveled full stroke then it will show open or close on re-discovery
    def setOn1(self, command):
        input = self.mapping[command['cmd']]['input']
        output = self.mapping[command['cmd']]['output']
        index = self.mapping[command['cmd']]['index']
        if self.bc.binaryOutput(index) == 0:
            self.bc.binaryOutput(index, 1)
            self.setDriver(input, 1)
            self.setDriver(output, 255)
            LOGGER.info('Output {} On'.format(index))
            time.sleep(.5)
        if self.bc.binaryOutput(index) != 0:
            self.bc.binaryOutput(index, 0)
            self.setDriver(input, 0)
            self.setDriver(output, 0)
            LOGGER.info('Output {} Off'.format(index))
            LOGGER.info('Door Operating')
```
I need to work on the delayed status as it is all if statements within setOn, some inconsistency after the delay.

```
        if self.bc.universalInput(index) == 12:
            self.setDrivers(output, '12', report=True, force=True)
        if self.bc.universalInput(index) == 5:
            self.setDriver(output, '5', report=True, force=True)
        if self.bc.universalInput(index) == 0:
            self.setDriver(output, '0', report=True, force=True)
            LOGGER.info('Door Operation Complete')
```


I added Bump back in so the door can be stopped anywhere in its travel it show STOPPED and if full travel then waits for a re-fresh.

Question, Bob what if the API you are using for your OWN poly had weather alerts like tornado, hurricane, earthquakes,
 how would you handle that? Do you still look at it on a refresh or would you constantly look at it looping in your Node Server?

The BASpi is its own little server with an internal API queried by the bascontrolns module with GET inputs and GET/SET for outputs with requests via its IP address. 
Now it is dynamic internally for the node readings but will not send anything until asked, all changeable.

I am cleaning this up today and adding a little to the readme so for the most part, close enough! 

Thank You for the PUSH and all of YOUR Valuable time and help!!!
Highest Regards!

**Comment by sjpbailey** [comment]
@bpaauwe !

 The problem with status having inconsistent status is that I was reading an input that was not the corrected from the raw value in ohms. 
 So I was not reading 12, or 5 that I was doing the if = to statement, I was reading 12,000+ and 5000+ for ohms.
 So I had to do a correction in the status of the inputs like this.
`if self.bc.universalInput(index)//1000 == 5:` in stead of `if self.bc.universalInput(index) == 5:` 

```
           if self.bc.universalInput(index)//1000 == 5:
                self.setDriver(output, '5', report=True, force=True)
                LOGGER.info('Door Open Complete')
            elif self.bc.universalInput(index)//1000 == 0:
                self.setDriver(output, '0', report=True, force=True)
                LOGGER.info('Door Stopped Complete')
            elif self.bc.universalInput(index)//1000 == 12:
                self.setDriver(output, '12', report=True, force=True)
                LOGGER.info('Door Closed Complete')
            else:
                return
```

**Comment by sjpbailey** [comment]
Hello Bob,

 Very detailed explanationI really appreciate your time! Understood about the template, it is not that I believed everything should be used it is my lack of Python experience. 
You have enlightened me to do better and to try to be a Python Developer which I am far from being.

I have a new video showing it working now. All due to one missing conversion to an input Node, https://youtu.be/p6vGUPnlAqs <https://youtu.be/p6vGUPnlAqs>

Now I can finish cleaning up my Irrigation Poly and finish my Fan Coil / Thermostat that uses an internal Sedona program running in the BASpi.

I will fix the refresh in config and remove the cost in the json.

I am so very pleased that you took the time to help me I have learned much in a short period of time. 
I did start writing actual code until 2019.
I am pleased with myself and wish my father was here to show because he first showed me TTL and MOSFIT electronics in the 80’s and my friend Gene Stopp building synthesizers in the 70’s.

**Comment by sjpbailey** [comment]
I will load this from the store then do a final condensing of the outputs like the converted inputs.
Will close issues after some more testing!

Again I cannot Thank You Enough for the Boost in knowledge to an old dog, rookie!

**Comment by bpaauwe** [comment]
It looks a lot better now.  Thanks for taking the comments in the spirit in which they were given.  I approved it for inclusion in the store.

**Comment by sjpbailey** [comment]
Thank You Again For This!

Now I have a better idea of how to condense my code!

Very important!
